### PR TITLE
ARROW-3677: [Go] Add fixed-length binary builder and array

### PR DESCRIPTION
**Pull request by alexandreyc**

### CODE REVIEW

File: go/arrow/array/binary_test.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package array
```
Comment (sbinet) [inline go/arrow/array/binary_test.go]: missing copyright stanza

File: go/arrow/array/fixedsize_binary.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package array
```
Comment (sbinet) [inline go/arrow/array/fixedsize_binary.go]: missing copyright stanza

File: go/arrow/array/fixedsize_binary_test.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+package array
```
Comment (sbinet) [inline go/arrow/array/fixedsize_binary_test.go]: missing copyright stanza

File: go/arrow/array/fixedsize_binarybuilder.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
+package array
```
Comment (sbinet) [inline go/arrow/array/fixedsize_binarybuilder.go]: missing copyright stanza

File: go/arrow/datatype_fixedwidth.go [hunk 5]
```diff
@@ -24,6 +24,14 @@ func (t *BooleanType) Name() string { return "bool" }
 // BitWidth returns the number of bits required to store a single element of this data type in memory.
 func (t *BooleanType) BitWidth() int { return 1 }
 
+type FixedSizeBinaryType struct {
```
Comment (sbinet) [inline go/arrow/datatype_fixedwidth.go]: we can't add this type to the `FixedWidthTypes` struct (well, we could, but it wouldn't make much sense.)
but perhaps we could test `FixedSizeBinaryType` implements the `FixedWidthDataType` interface using the usual `_ FixedWidthDataType = (*FixedSizeBinaryType)(nil)` trick?

File: go/arrow/array/fixedsize_binary_test.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/apache/arrow/go/arrow"
+	"github.com/apache/arrow/go/arrow/memory"
+)
+
+func TestFixedSizeBinary(t *testing.T) {
+	mem := memory.NewCheckedAllocator(memory.NewGoAllocator())
+	defer mem.AssertSize(t, 0)
+
+	dtype := arrow.FixedSizeBinaryType{ByteWidth: 7}
+	b := NewFixedSizeBinaryBuilder(mem, dtype)
```
Comment (sbinet) [inline go/arrow/array/fixedsize_binary_test.go]: perhaps add a "dummy" pair of `Retain()/Release()` to exercize that API ?

File: go/arrow/array/fixedsize_binary_test.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+	values := [][]byte{
+		[]byte("7654321"),
+		nil,
+		[]byte("AZERTYU"),
+	}
+	b.AppendValues(values, []bool{true, false, true})
+
+	a := b.NewFixedSizeBinaryArray()
```
Comment (sbinet) [inline go/arrow/array/fixedsize_binary_test.go]: ditto: perhaps add a "dummy" pair of `Retain/Release` ?

File: go/arrow/array/fixedsize_binarybuilder.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+	for _, vv := range v {
+		b.appendNextOffset()
+		b.values.Append(vv)
+	}
+
+	b.builder.unsafeAppendBoolsToBitmap(valid, len(v))
+}
+
+func (b *FixedSizeBinaryBuilder) Value(i int) []byte {
+	// start := i * b.dtype.ByteWidth
```
Comment (sbinet) [inline go/arrow/array/fixedsize_binarybuilder.go]: please remove.

File: go/arrow/array/fixedsize_binarybuilder.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+// by checking the capacity and calling Resize if necessary.
+func (b *FixedSizeBinaryBuilder) Reserve(n int) {
+	b.builder.reserve(n, b.Resize)
+}
+
+// Resize adjusts the space allocated by b to n elements. If n is greater than b.Cap(),
+// additional memory will be allocated. If n is smaller, the allocated memory may reduced.
+func (b *FixedSizeBinaryBuilder) Resize(n int) {
+	b.offsets.resize((n + 1) * arrow.Int32SizeBytes)
+	//b.values.resize(n * b.dtype.ByteWidth)
```
Comment (sbinet) [inline go/arrow/array/fixedsize_binarybuilder.go]: please remove.

File: go/arrow/array/fixedsize_binary_test.go (hunk 7 above)
Comment (sbinet) [reply go/arrow/array/fixedsize_binary_test.go]: also: perhaps use the `NewArray` API (with a type-cast to `*FixedSizeBinary`) to also make sure `NewArray` works correctly.

File: go/arrow/array/fixedsize_binary.go [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+// A type which represents an immutable sequence of fixed-length binary strings.
+type FixedSizeBinary struct {
+	array
+	byteWidth    int
+	valueOffsets []int32
+	valueBytes   []byte
+}
+
+// NewFixedSizeBinaryData constructs a new fixed-size binary array from data.
+func NewFixedSizeBinaryData(data *Data) *FixedSizeBinary {
```
Comment (sbinet) [inline go/arrow/array/fixedsize_binary.go]: this needs to be wired into `makeArrayFn`:
https://github.com/apache/arrow/blob/0a8654d3060e9f83a33e8bf22984b80e08cadffe/go/arrow/array/array.go#L182

(together with a test "of course" :))

File: go/arrow/array/fixedsize_binary.go [hunk 11]
```diff
@@ -0,0 +1,45 @@
+package array
+
+import "github.com/apache/arrow/go/arrow"
+
+// A type which represents an immutable sequence of fixed-length binary strings.
+type FixedSizeBinary struct {
```
Comment (sbinet) [inline go/arrow/array/fixedsize_binary.go]: perhaps add the usual trick to test that `FixedSizeBinary` implements `array.Interface` ?

**Review by sbinet — COMMENTED** [review]
this is just a first quick pass.
I'll come back tomorrow.

**Comment by alexandreyc** [comment]
Should be good now.

I got ride of attribute `byteWidth` on struct `FixedSizeBinary` because I realized it was useless when writing the test for array api.

-Alexandre

**Comment by alexandreyc** [comment]
Fixed an inconsistency and added missing cases to newBuidler.

**Review by sbinet — APPROVED** [review]
LGTM

**Pull request closed without merging**